cloudinit/config/cc_grub_dpkg.py

```diff
@@ -43,10 +44,58 @@
 import os
 
 from cloudinit import util
+from cloudinit.util import ProcessExecutionError
 
 distros = ['ubuntu', 'debian']
 
 
+def fetch_idevs(log):
+    """
+    Fetches the /dev/disk/by-id device grub is installed to.
+    Falls back to plain disk name if no by-id entry is present.
+    """
+    disk = ""
+    devices = []
+
+    try:
+        # get the root disk where the /boot directory resides.
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
+                         capture=True)[0].strip()
+    except ProcessExecutionError as e:
+        # grub-common may not be installed, especially on containers
+        # FileNotFoundError is a nested exception of ProcessExecutionError
+        if isinstance(e.reason, FileNotFoundError):
+            log.debug("grub-common is not installed, e.g. inside a container")
+        # disks from the container host are present in /proc and /sys
+        # which is where grub-probe determines where /boot is.
+        # it then checks for existence in /dev, which fails as host disks
+        # are not exposed to the container.
+        elif "failed to get canonical path" in e.stderr:
+            log.debug("Device mapping between /proc/self/mountinfo does not "
+                      "match /dev, e.g. inside a container")
+        else:
+            # something bad has happened, continue to log the error
+            raise
+    except Exception:
+        util.logexc(log, "grub-probe failed to execute for grub-dpkg")
+
+    try:
+        # check if disk exists and use udevadm to fetch symlinks
+        if os.path.exists(disk):
```

Comment:
This relies on the `os.path.exists("")` returning `False`; this seems like a reasonable assumption, but just in case there are unanticipated corner cases, we could be more explicit:

```suggestion
        if disk and os.path.exists(disk):
```